### PR TITLE
feat: Add ability to configure loglevel for each ArgoCD component

### LINKED ISSUE #329: Setting loglevel for components
**Is your feature request related to a problem? Please describe.**
When using ArgoCD operator the loglevel of all components is set to the default value of 'INFO'. Depending on the use case this will produce a lot of unrelated log messages. As far as I see there is no option to adjust the loglevel of the ArgoCD components.

**Describe the solution you'd like**
Using the ArgoCD CR we could set the loglevel for the different components.

**Describe alternatives you've considered**
Setting the --loglevel in the Deployment will not work as these resources are operator managed.

**Additional context**
The loglevel could be set for all components, e.g. https://argoproj.github.io/argo-cd/operator-manual/server-commands/argocd-server/.

### PR DESCRIPTION
**Pull request by tylerauerbeck**
**What type of PR is this?**

> /kind enhancement

**What does this PR do / why we need it**:
This PR adds the ability to adjust the loglevel of each of the ArgoCD components (controller, repo, server)

**Have you updated the necessary documentation?**

* [X] Documentation update is required by this PR.
* [X] Documentation has been updated.

**Which issue(s) this PR fixes**:

Fixes #329 

**How to test changes / Special notes to the reviewer**:

After building the operator from this PR branch, you can deploy it with the following to confirm the loglevel functionality:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  name: example-argocd
  labels:
    example: basic
spec: 
  server:
    logLevel: "warn"
  repo:
    logLevel: "error"
  controller:
    logLevel: "info"
```

If you don't set any of the above, it will default the loglevel to info (which is the current behavior of each of the components).

### CODE REVIEW

File: docs/reference/argocd.md [hunk 1]
```diff
@@ -123,6 +123,7 @@ Name | Default | Description
 Processors.Operation | 10 | The number of operation processors.
 Processors.Status | 20 | The number of status processors.
 Resources | [Empty] | The container compute resources.
+LogLevel | info | The log level to be used by the ArgoCd Application Controller component.
```
Comment (wtam2018) [inline docs/reference/argocd.md:127]: Please include accepted loglevels: `info`, `warn`, `error`, `debug`.

File: pkg/apis/argoproj/v1alpha1/argocd_types.go [hunk 2]
```diff
@@ -58,6 +58,9 @@ type ArgoCDApplicationControllerSpec struct {
 	// Processors contains the options for the Application Controller processors.
 	Processors ArgoCDApplicationControllerProcessorsSpec `json:"processors,omitempty"`
 
+	// LogLevel refers to the log level used by the Application Controller component. Defaults to ArgoCDDefaultLogLevel if not configured.
```
Comment (wtam2018) [inline pkg/apis/argoproj/v1alpha1/argocd_types.go:61]: include the accepted loglevels (in the generated description) is useful for the users.

File: pkg/controller/argocd/deployment.go [hunk 3]
```diff
@@ -194,6 +194,10 @@ func getArgoRepoCommand(cr *argoprojv1a1.ArgoCD) []string {
 	cmd = append(cmd, "--redis")
 	cmd = append(cmd, getRedisServerAddress(cr))
 
+	cmd = append(cmd, "--loglevel "+getLogLevel(cr.Spec.Repo.LogLevel))
+
+	//cmd = append(cmd, getLogLevel(cr.Spec.Repo.LogLevel))
```
Comment (wtam2018) [inline pkg/controller/argocd/deployment.go:199]: please remove this comment.

File: pkg/controller/argocd/util.go [hunk 4]
```diff
@@ -1204,3 +1207,20 @@ func initK8sClient() (*kubernetes.Clientset, error) {
...
+func getLogLevel(logField string) string {
+
+	if logField == "" {
+		return common.ArgoCDDefaultLogLevel
+	} else {
+		switch logField {
+		case
+			"info",
+			"warn",
+			"error":
```
Comment (wtam2018) [inline pkg/controller/argocd/util.go:1221]: Is `debug` a supported loglevel?

**Review by wtam2018 — COMMENTED** [review]
It looks good.   Just a few questions.   Thanks.

File: pkg/controller/argocd/util.go (hunk 4 above)
Comment (tylerauerbeck) [reply pkg/controller/argocd/util.go:1221]: According to https://argoproj.github.io/argo-cd/operator-manual/server-commands/argocd-server/ (and applies to each of the components), it looks like the only options are `info`,`warn`, and `error`

File: pkg/apis/argoproj/v1alpha1/argocd_types.go (hunk 2 above)
Comment (tylerauerbeck) [reply pkg/apis/argoproj/v1alpha1/argocd_types.go]: Done

**Comment by tylerauerbeck** [comment]
@wtam2018 Updated based on your requests. But I do see that it appears that the controller and server pods are entering `CrashLoopBackOff`. Going to take a look at this now.

File: pkg/controller/argocd/util.go (hunk 4 above)
Comment (wtam2018) [reply pkg/controller/argocd/util.go:1221]: I see this from the reference you provided.   Please double check
```
      --loglevel string                               Set the logging level. One of: debug|info|warn|error (default "info")
```

**Comment by wtam2018** [comment]
Any chance we can add this set loglevel feature to applicationset-controller as well?  :-)

File: pkg/controller/argocd/util.go (hunk 4 above)
Comment (tylerauerbeck) [reply pkg/controller/argocd/util.go:1223]: Ah. Must have overlooked this. This is fixed now.

**Comment by tylerauerbeck** [comment]
@wtam2018 Think I've got everything fixed up now. I've also included the loglevel for applicationset. Just waiting on the tests to pass now.

**Comment by tylerauerbeck** [comment]
@wtam2018 Looks like all is good here. Are there any other changes you wanted to see?

File: docs/reference/argocd.md [hunk 5]
```diff
@@ -123,6 +123,7 @@ Name | Default | Description
 Processors.Operation | 10 | The number of operation processors.
 Processors.Status | 20 | The number of status processors.
 Resources | [Empty] | The container compute resources.
+LogLevel | info | The log level to be used by the ArgoCd Application Controller component. Valid options are info, error, and warn.
```
Comment (wtam2018) [inline docs/reference/argocd.md:127]:
```suggestion
LogLevel | info | The log level to be used by the ArgoCd Application Controller component. Valid options are debug, info, error, and warn.

```

File: docs/reference/argocd.md [hunk 6]
```diff
@@ -769,6 +770,7 @@ VerifyTLS | false | Whether to enforce strict TLS checking on all components whe
 AutoTLS | "" | Provider to use for setting up TLS the repo-server's gRPC TLS certificate (one of: `openshift`). Currently only available for OpenShift.
 Image | `argoproj/argocd` | The container image for ArgoCD Repo Server. This overrides the `ARGOCD_REPOSERVER_IMAGE` environment variable.
 Version | v2.0.0 (SHA) | The tag to use with the ArgoCD Repo Server.
+LogLevel | info | The log level to be used by the ArgoCD Repo Server. Valid options are info, error, and warn.
```
Comment (wtam2018) [inline docs/reference/argocd.md:774]:
```suggestion
LogLevel | info | The log level to be used by the ArgoCD Repo Server. Valid options are debug, info, error, and warn.

```

File: docs/reference/argocd.md [hunk 7]
```diff
@@ -923,6 +925,7 @@ Insecure | false | Toggles the insecure flag for Argo CD Server.
 Resources | [Empty] | The container compute resources.
 [Route](#server-route-options) | [Object] | Route configuration options.
 Service.Type | ClusterIP | The ServiceType to use for the Service resource.
+LogLevel | info | The log level to be used by the ArgoCD Server component. Valid options are info, error, and warn.
```
Comment (wtam2018) [inline docs/reference/argocd.md:929]:
```suggestion
LogLevel | info | The log level to be used by the ArgoCD Server component. Valid options are debug, info, error, and warn.

```

File: pkg/apis/argoproj/v1alpha1/argocd_types.go [hunk 8]
```diff
@@ -58,6 +58,9 @@ type ArgoCDApplicationControllerSpec struct {
 	// Processors contains the options for the Application Controller processors.
 	Processors ArgoCDApplicationControllerProcessorsSpec `json:"processors,omitempty"`
 
+	// LogLevel refers to the log level used by the Application Controller component. Defaults to ArgoCDDefaultLogLevel if not configured. Valid options are debug,info, error, and warn.
+	LogLevel string `json:"logLevel,omitempty"`
```
Comment (wtam2018) [inline pkg/apis/argoproj/v1alpha1/argocd_types.go:62]:
```suggestion
	// LogLevel refers to the log level used by the Application Controller component. Defaults to ArgoCDDefaultLogLevel if not configured. Valid options are debug, info, error, and warn.
	LogLevel string `json:"logLevel,omitempty"`
```

File: pkg/controller/argocd/util.go [hunk 9]
```diff
@@ -1204,3 +1208,21 @@ func initK8sClient() (*kubernetes.Clientset, error) {
...
+		switch logField {
+		case
+			"debug",
+			"info",
+			"warn",
+			"error":
+			return logField
+		}
+		return common.ArgoCDDefaultLogLevel
+	}
```
Comment (wtam2018) [inline pkg/controller/argocd/util.go:1227]:
```suggestion
        switch strings.ToLower(logField) {
	case "debug",
		"info",
		"warn",
		"error":
		return logField
	}
	return common.ArgoCDDefaultLogLevel
```

**Review by wtam2018 — COMMENTED** [review]
Thanks for updating.   It looks good.  Have a few minor suggestions.   (note:  indent is a bit off)

File: pkg/controller/argocd/util.go [hunk 10]
```diff
@@ -1204,3 +1208,16 @@ func initK8sClient() (*kubernetes.Clientset, error) {
 
 	return k8sClient, nil
 }
+
+// getLogLevel returns the log level for a specified component if it is set or returns the default log level if it is not set
+func getLogLevel(logField string) string {
+
+        switch strings.ToLower(logField) {
+	case "debug",
```
Comment (wtam2018) [inline pkg/controller/argocd/util.go:1216]:
```suggestion
	switch strings.ToLower(logField) {
	case "debug",
```

File: deploy/crds/argoproj.io_argocds_crd.yaml [hunk 11]
```diff
@@ -42,6 +42,11 @@ spec:
                   image:
                     description: Image is the Argo CD ApplicationSet image (optional)
                     type: string
+                  logLevel:
```
Comment (wtam2018) [inline deploy/crds/argoproj.io_argocds_crd.yaml:45]: Please re-generate CRD.

Comment (tylerauerbeck) [reply deploy/crds/argoproj.io_argocds_crd.yaml:45]: Done.

**Review by wtam2018 — APPROVED** [review]
LGTM
thanks @tylerauerbeck !

**Pull request merged**